### PR TITLE
Starting work on ballot voting

### PR DESCRIPTION
**Pull request by C4L86**

### CODE REVIEW

File: app/controllers/nominations_controller.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ def nomination_ballot
 
   def save_nominee
     current_user.record_nominations(current_user.id, params[:awards])
+
```
Comment (sumeetjain) [inline app/controllers/nominations_controller.rb:10]: Remove trailing whitespace

File: app/views/nominations/index.html.erb [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
-  <ul>
-  <% sorted_award_noms = a.nominations.order(:nominee) %>
-  <% sorted_award_noms.each do |nom| %> 
-   <li><%= nom.nominee %>, Weight: <%= nom.weight %> </li>
-  <% end %>
-  </ul>
+    <%= a.award_name %>
+    <ul>
+    <% sorted_award_noms = a.nominations.order(:nominee) %>
+      <% sorted_award_noms.each do |nom| %>
```
Comment (sumeetjain) [inline app/views/nominations/index.html.erb:7]: Remove trailing whitespace

File: app/views/votes/ballots_page.html.erb [hunk 3]
```diff
@@ -10,12 +10,12 @@
...
-            Role: <%= item.role %> <br> 
-            Play: <%= item.play.title %> <br> 
-            Theater: <%= item.play.theater.name %> <br>
-            <% end %>
-          <% end %>
+<% @awards.each do |award| %>
+ <h2> <%= award.award_name %> </h2>
+      <% award.ballot_items.each do |item| %>
+        Nominee: <%= item.nominee %> <br>
+        Role: <%= item.role %> <br>
```
Comment (sumeetjain) [inline app/views/votes/ballots_page.html.erb]: Remove trailing whitespace

File: app/controllers/nominations_controller.rb (hunk 1 above)
Comment (sumeetjain) [inline app/controllers/nominations_controller.rb:10]: Trailing whitespace detected.

File: app/dashboards/nomination_dashboard.rb [hunk 4]
```diff
@@ -32,6 +32,7 @@ class NominationDashboard < Administrate::BaseDashboard
     :award,
     :show,
     :weight,
+    :approved,
```
Comment (sumeetjain) [inline app/dashboards/nomination_dashboard.rb:35]: Avoid comma after the last item of an array.

File: app/models/user.rb [hunk 5]
```diff
@@ -69,6 +69,8 @@ def save_nomination_object(user_id, current_award, value3)
     @new_nom.user_id = user_id
     @new_nom.award_id = current_award
     @new_nom.open = true
+    # Chris added below 'approved' line for Admin to change to 'true' when it passes their check. If 'true', the nomination will appear in the ballot_box for user vote
```
Comment (sumeetjain) [inline app/models/user.rb:72]: Line is too long. [96/80]

Comment (sumeetjain) [inline app/models/user.rb:72]: Line is too long. [167/80]

File: app/views/votes/ballots_page.html.erb [hunk 6]
```diff
@@ -9,13 +9,13 @@
...
-            Role: <%= item.role %> <br> 
-            Play: <%= item.play.title %> <br> 
-            Theater: <%= item.play.theater.name %> <br>
-            <% end %>
-          <% end %>
+<% @nominations.each do |nom| %>
+  <% if nom.approved == true %>
+   <h2> <%= nom.award_id %> </h2>
+          Nominee: <%= nom.nominee %> <br>
+          Role: <%= nom.role %> <br>
```
Comment (sumeetjain) [inline app/views/votes/ballots_page.html.erb]: Remove trailing whitespace

File: app/controllers/votes_controller.rb [hunk 7]
```diff
@@ -5,7 +5,10 @@ class VotesController < ApplicationController
   def ballots_page
     @all_ballot_items = BallotItem.all 
     # Really you'll want something like BallotItem.all.awards in case of rogue award categories.  
-    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that. 
+    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that.
```
Comment (sumeetjain) [inline app/controllers/votes_controller.rb:8]: Line is too long. [138/80]

File: app/controllers/votes_controller.rb [hunk 8]
```diff
@@ -5,7 +5,10 @@ class VotesController < ApplicationController
...
+
+    # I think you'll want to use the nominations model for this and reference awards via ActiveRecord relations
```
Comment (sumeetjain) [inline app/controllers/votes_controller.rb]: Line is too long. [111/80]

File: app/controllers/votes_controller.rb [hunk 9]
```diff
@@ -5,7 +5,10 @@ class VotesController < ApplicationController
   def ballots_page
     @all_ballot_items = BallotItem.all 
     # Really you'll want something like BallotItem.all.awards in case of rogue award categories.  
-    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that. 
+    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that.
+
+    # I think you'll want to use the nomination model for this and reference awards via ActiveRecord relations
```
Comment (sumeetjain) [inline app/controllers/votes_controller.rb]: Line is too long. [110/80]

File: app/controllers/admin/ballot_items_controller.rb [hunk 10]
```diff
@@ -15,5 +15,8 @@ class BallotItemsController < Admin::ApplicationController
 
     # See https://administrate-docs.herokuapp.com/customizing_controller_actions
     # for more information
+    def ballots_page
```
Comment (sumeetjain) [inline app/controllers/admin/ballot_items_controller.rb]: Remove unused methods (admin::ballotitemscontroller#ballots_page)

File: app/models/ballot_item.rb [hunk 11]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
 belongs_to :award
 belongs_to :play
 
+# play_id == convergence of play title and theater
+  def approved_ballot
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Remove unused methods (ballotitem#approved_ballot)

File: app/models/ballot_item.rb [hunk 12]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+
+    approved_nominations.each do |nom|
+      ballot.nominee  = nom.nominee
+      ballot.role     = nom.role
+      ballot.award_id = nom.award_id
+      ballot.play_id  = Play.where("theater == ? AND title == ?", nom.theater, nom.show).id
+
+      ballot.save
+    end
+  end
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Remove trailing whitespace

File: app/models/ballot_item.rb (hunk 11 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Has approx 8 statements (TooManyStatements)

Comment (sumeetjain) [inline app/models/ballot_item.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: app/models/ballot_item.rb [hunk 13]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
 belongs_to :award
 belongs_to :play
 
+# play_id == convergence of play title and theater
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Incorrect indentation detected (column 0 instead of 2).

File: app/models/ballot_item.rb (hunk 11 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Inconsistent indentation detected.

File: app/models/ballot_item.rb [hunk 14]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+  def approved_ballot
+    ballot = BallotItem.new
+    approved_nominations = Nomination.where({"approved" => true})
+
+    approved_nominations.each do |nom|
+      ballot.nominee  = nom.nominee
+      ballot.role     = nom.role
+      ballot.award_id = nom.award_id
+      ballot.play_id  = Play.where("theater == ? AND title == ?", nom.theater, nom.show).id
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Line is too long. [91/80]

File: app/models/ballot_item.rb (hunk 12 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Trailing whitespace detected.

File: app/models/ballot_item.rb [hunk 15]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+
+      ballot.save
+    end
+  end 
+
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Extra empty line detected at class body end.

File: app/controllers/votes_controller.rb [hunk 16]
```diff
@@ -5,8 +5,9 @@ class VotesController < ApplicationController
   def ballots_page
     @all_ballot_items = BallotItem.all 
     # Really you'll want something like BallotItem.all.awards in case of rogue award categories.  
-    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that. 
-  end
-
+    @awards = Award.all # This probably won't work in case there is a rogue ballot item that does not have an award_id.....TODO: Fix that.
```
Comment (sumeetjain) [inline app/controllers/votes_controller.rb:8]: Line is too long. [138/80]

File: app/controllers/votes_controller.rb [hunk 17]
```diff
@@ -5,8 +5,9 @@ class VotesController < ApplicationController
...
 
+
```
Comment (sumeetjain) [inline app/controllers/votes_controller.rb]: Trailing whitespace detected.

File: app/models/ballot_item.rb [hunk 18]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+# play_id == convergence of play title and theater
+  def self.approved_ballots
+    ballot = BallotItem.new
+    approved_nominations = Nomination.where({"approved" => true})
+
+    approved_nominations.each do |nom|
+      ballot.nominee  = nom.nominee
+      ballot.role     = nom.role
+      ballot.award_id = nom.award_id
+      ballot.play_id  = Play.where("theater_id == ? AND title == ?", Theater.find_by_name(nom.theater).id, nom.show).ids
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Line is too long. [120/80]

File: app/models/ballot_item.rb [hunk 19]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
 belongs_to :award
 belongs_to :play
 
+
+
+  def self.top_ballots
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Remove unused methods (ballotitem#top_ballots)

File: app/models/nomination.rb [hunk 20]
```diff
@@ -24,5 +24,11 @@ def self.rank_by_voter_weight
     return sorted_noms
   end
 
+  def self.approved_nominations
+    self.where({"approved" => true})
+  end
 
+  def ranked_nominations
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Remove unused methods (nomination#ranked_nominations)

File: app/models/ballot_item.rb [hunk 21]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+    ballot = BallotItem.new
+    approved_nominations =
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Remove trailing whitespace

File: app/models/nomination.rb [hunk 22]
```diff
@@ -24,5 +24,11 @@ def self.rank_by_voter_weight
...
+
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Remove trailing whitespace

File: app/models/ballot_item.rb (hunk 19 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Has approx 7 statements (TooManyStatements)

File: app/models/ballot_item.rb [hunk 23]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
 belongs_to :award
 belongs_to :play
 
+
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Extra blank line detected.

File: app/models/ballot_item.rb [hunk 24]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Extra blank line detected.

File: app/models/ballot_item.rb (hunk 19 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Inconsistent indentation detected.

File: app/models/ballot_item.rb (hunk 21 above)
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Useless assignment to variable - `approved_nominations`.

Comment (sumeetjain) [inline app/models/ballot_item.rb]: Trailing whitespace detected.

File: app/models/ballot_item.rb [hunk 25]
```diff
@@ -3,4 +3,19 @@ class BallotItem < ActiveRecord::Base
...
+  def self.top_ballots
+    ballot = BallotItem.new
+    approved_nominations = 
+
+    approved_nominations.each do |nom|
```
Comment (sumeetjain) [inline app/models/ballot_item.rb]: Indent the first line of the right-hand-side of a multi-line assignment.

File: app/models/nomination.rb (hunk 22 above)
Comment (sumeetjain) [inline app/models/nomination.rb]: Trailing whitespace detected.

File: app/models/nomination.rb [hunk 26]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
...
+    identical_noms = []
+    final_noms = []
+
+    approved_noms.each do |nom|
+      identical_noms << approved_noms.where("nominee == ? AND role == ? AND award_id == ? AND theater == ? AND show == ?", nom.nominee, nom.role, nom.award_id, nom.theater, nom.show)
+    end
+    identical_noms
+  end
+
+  def self.duplicate_noms
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Remove unused methods (nomination#duplicate_noms)

File: app/models/nomination.rb [hunk 27]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
     return sorted_noms
   end
 
+  # def self.approved_nominations
+  #   self.where({"approved" => true})
+  # end
 
+  def self.ranked_nominations
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Has approx 6 statements (TooManyStatements)

File: app/models/nomination.rb [hunk 28]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
...
+    approved_noms = self.where({"approved" => true})
+    identical_noms = []
+    final_noms = []
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Useless assignment to variable - `final_noms`.

File: app/models/nomination.rb [hunk 29]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
...
+
+    approved_noms.each do |nom|
+      identical_noms << approved_noms.where("nominee == ? AND role == ? AND award_id == ? AND theater == ? AND show == ?", nom.nominee, nom.role, nom.award_id, nom.theater, nom.show)
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [182/80]

File: app/models/nomination.rb [hunk 30]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
...
+    end
+    identical_noms
+  end
+
+  def self.duplicate_noms
+    self.where({"approved" => true}).select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).having("count(*) > 1")
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Space missing after comma.

Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [154/80]

File: app/models/nomination.rb [hunk 31]
```diff
@@ -24,5 +24,22 @@ def self.rank_by_voter_weight
...
+  #   self.where({"approved" => true})
+  # end
 
+  def self.ranked_nominations
+    approved_noms = self.where({"approved" => true})
+    identical_noms = {}
+    final_noms = []
+
+    approved_noms.each do |nom|
+      identical_noms[nom.id] = approved_noms.where("nominee == ? AND role == ? AND award_id == ? AND theater == ? AND show == ?", nom.nominee, nom.role, nom.award_id, nom.theater, nom.show)
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [189/80]

**Comment by C4L86** [comment]
The Nomination class method that I've defined, `duplicate_noms`, is able to recognize approved nominations sharing identical Nomination attributes for nominee, role, award_id, theater, and show. From this return, I am hoping to create a rank based on how many times each duplicate occurs each weighted based on the user's weight who made the nomination. However, I am having trouble taking what is returned and identifying the count of the duplicates as well as the user_ids associated with each. Ideas?

**Comment by sumeetjain** [comment]
I would approach this backwards:

What data do you need to build the Nominations Report page? How would it be structured?

I would guess that looks something like:

``` ruby
{
  # 2 here refers to award #2
  2:  {
          {theater: "Blah", show: "Foo"} => 35, # meaning this show has 35 votes, after all weighting is applied.
          {theater: "Blah Blah", show: "Foo Foo"} => 31,
          {theater: "a", show: "b"} => 29,
          {theater: "c", show: "e"} => 10,
          {theater: "b", show: "l"} => 9,
          {theater: "p", show: "n"} => 5,
          {theater: "z", show: "x"} => 3,
          {theater: "q", show: "m"} => 3,
          {theater: "r", show: "y"} => 1
        }

   5: {
      # ...
   }
}
```

Do whatever looping and collecting you need to do to get this structure.

**Comment by sumeetjain** [comment]
Let me know if that's not helpful.

**Comment by C4L86** [comment]
I guess I'm not seeing how this is taking a backwards approach. From my understanding, any given award can yield an unanticipated nomination as entered by each user. That's why I want to identify any given nomination, as approved by the admin, and look for duplicates as those would most likely have more weight than any single nomination. Could I talk with you in person?

**Comment by sumeetjain** [comment]
For sure–come on over.

**Comment by C4L86** [comment]
I'm running into an ActiveRecord Configuration error related to the association between the nominations and users tables and am unable to remedy the issue. Here's a screenshot of the error:
screen shot 2016-04-07 at 2 41 29 pm
Otherwise, I feel like I'm getting very close to the solution for the ballots.

File: app/models/nomination.rb [hunk 32]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
+  def self.duplicate_noms
+    self.approved_nominations.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
+  def self.final_hash
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Remove unused methods (nomination#final_hash)

File: app/models/nomination.rb [hunk 33]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+    x = {}
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Has the variable name 'x' (UncommunicativeVariableName)

File: app/models/nomination.rb [hunk 34]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
   end
 
+  # NOT BEING USED. JUST TESTING
+  # def self.ranked_nominations
+  #   approved_noms = self.where({"approved" => true})
+  #   identical_noms = {}
+  #   final_noms = []
+
+  #   approved_noms.each do |nom|
+  #     identical_noms[nom.id] = approved_noms.where("nominee == ? AND role == ? AND award_id == ? AND theater == ? AND show == ?", nom.nominee, nom.role, nom.award_id, nom.theater, nom.show)
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [191/80]

File: app/models/nomination.rb [hunk 35]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+  #   end
+  #   identical_noms
+  # end
+
+  # Defines an 'approved nomination' as one having its approved attribute set to 'true' by the Admin
```
Comment (sumeetjain) [inline app/models/nomination.rb:27]: Line is too long. [100/80]

File: app/models/nomination.rb [hunk 36]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+  def self.approved_nominations
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
```
Comment (sumeetjain) [inline app/models/nomination.rb:32]: Line is too long. [96/80]

File: app/models/nomination.rb [hunk 37]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+  def self.duplicate_noms
+    self.approved_nominations.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Space missing after comma.

Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [175/80]

File: app/models/nomination.rb [hunk 38]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
```
Comment (sumeetjain) [inline app/models/nomination.rb:37]: Line is too long. [259/80]

File: app/models/nomination.rb [hunk 39]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+  def self.final_hash
+    x = {}
+
+    self.duplicate_noms.each do |nom|
+      # array of all of the weights of all users who submitted nominations for this nominee
```
Comment (sumeetjain) [inline app/models/nomination.rb:42]: Line is too long. [91/80]

File: app/models/nomination.rb [hunk 40]
```diff
@@ -24,5 +24,41 @@ def self.rank_by_voter_weight
...
+      user_weights = Nomination.where(nominee: nom.nominee, role: nom.role, award_id: nom.award_id, theater: nom.theater, show: nom.show).joins(:users).pluck("users.weight")
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [173/80]

**Comment by halfghaninne** [comment]
What line of code is throwing this error? Are you trying to call `.users` on a Nomination object?

**Comment by C4L86** [comment]
I am trying to identify the user tied to a Nomination object and identify that user's weight (from the users table). From working with Sumeet, it was my understanding that `joins` makes the connection between the nominations and users table to then identify that weight.

**Comment by halfghaninne** [comment]
I may be missing something, but that approach seems heavy-handed. Is something like `@nomination.user.weight` not what you are looking for?

**Comment by C4L86** [comment]
Well I need the user's weight who is associated with the particular nomination that they made

**Comment by sumeetjain** [comment]
I think it's hard to see the overall approach without a deeper context for this. The joins is nice here, because it saves us from potentially thousands of calls to `nomination.user.weight` (each of which would be a SQL hit).

Chris, I think I meant to have you do `joins(:user)` not `:users`.

**Comment by C4L86** [comment]
I had tried `:user` before and that lead to another error, which I suppose is another step forward.

**Comment by C4L86** [comment]
Aaaand I figured it out. The users which I had created for some sample data hadn't been given a weight. I think that we're almost there!

File: app/models/nomination.rb [hunk 41]
```diff
@@ -24,5 +24,42 @@ def self.rank_by_voter_weight
...
+    self.approved_nominations.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
+  def self.final_hash
+    x = {}
+
+    self.duplicate_noms.each do |nom|
+      # array of all of the weights of all users who submitted nominations for this nominee
+      user_weights = Nomination.where(nominee: nom.nominee, role: nom.role, award_id: nom.award_id, theater: nom.theater, show: nom.show).joins(:user).pluck("users.weight")
```
Comment (sumeetjain) [inline app/models/nomination.rb:43]: Line is too long. [172/80]

File: app/models/nomination.rb [hunk 42]
```diff
@@ -24,5 +24,42 @@ def self.rank_by_voter_weight
...
+
+
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Extra blank line detected.

File: app/models/nomination.rb [hunk 43]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
+  def self.duplicate_nominations
+    self.approved_nominations.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
+  def self.ranked_nominations
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Remove unused methods (nomination#ranked_nominations)

File: app/models/nomination.rb [hunk 44]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
     return sorted_noms
   end
 
+  # Defines an 'approved nomination' as one having its approved attribute set to 'true' by the Admin
```
Comment (sumeetjain) [inline app/models/nomination.rb:27]: Line is too long. [100/80]

File: app/models/nomination.rb [hunk 45]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+  def self.approved_nominations
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
+  def self.duplicate_nominations
+    self.approved_nominations.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
```
Comment (sumeetjain) [inline app/models/nomination.rb]: Space missing after comma.

Comment (sumeetjain) [inline app/models/nomination.rb]: Line is too long. [175/80]

File: app/models/nomination.rb [hunk 46]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
```
Comment (sumeetjain) [inline app/models/nomination.rb:37]: Line is too long. [259/80]

File: app/models/nomination.rb [hunk 47]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+  def self.ranked_nominations
+    ranks = {}
+
+    self.duplicate_nominations.each do |nom|
+      # array of all of the weights of all users who submitted nominations for this nominee
```
Comment (sumeetjain) [inline app/models/nomination.rb:42]: Line is too long. [91/80]

File: app/models/nomination.rb [hunk 48]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+      user_weights = Nomination.where(nominee: nom.nominee, role: nom.role, award_id: nom.award_id, theater: nom.theater, show: nom.show).joins(:user).pluck("users.weight")
```
Comment (sumeetjain) [inline app/models/nomination.rb:43]: Line is too long. [172/80]

File: app/models/nomination.rb [hunk 49]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
+  def self.duplicates
+    self.approved_by_admin.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
+  end
+
+  # Returns a hash with keys of single iterations of duplicate nominations (as found by the above method 'duplicate_noms') and values of that nomination's weight as defined by users' weights (who made the nominations) and how many times that nomination occurs
+  def self.ranked_by_weight
```
Comment (sumeetjain) [inline app/models/nomination.rb:38]: Remove unused methods (nomination#ranked_by_weight)

Comment (sumeetjain) [inline app/models/nomination.rb:38]: Has approx 6 statements (TooManyStatements)

File: app/models/nomination.rb [hunk 50]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+    ranks = {}
+
+    self.duplicates.each do |nom|
+      # array of all of the weights of all users who submitted nominations for this nominee
+      user_weights = Nomination.where(nominee: nom.nominee, role: nom.role, award_id: nom.award_id, theater: nom.theater, show: nom.show).joins(:user).pluck("users.weight")
+
+      # sum the array. that's the total weight for all of these nominations.
+      ranks[nom] = user_weights.sum * nom.count
+    end
+    ranks.sort_by{|k,v| v}.reverse.to_h
```
Comment (sumeetjain) [inline app/models/nomination.rb:48]: Has the variable name 'k' (UncommunicativeVariableName)

Comment (sumeetjain) [inline app/models/nomination.rb:48]: Has the variable name 'v' (UncommunicativeVariableName)

File: app/models/nomination.rb [hunk 51]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
     return sorted_noms
   end
 
+  # Defines an 'approved nomination' as one having its approved attribute set to 'true' by the Admin
+  def self.approved_by_admin
+    self.where({"approved" => true})
+  end
+
+  # Identifies nominations sharing identical attributes (nominee, role, award_id, theater, show)
```
Comment (sumeetjain) [inline app/models/nomination.rb:32]: Line is too long. [96/80]

File: app/models/nomination.rb [hunk 52]
```diff
@@ -24,5 +24,28 @@ def self.rank_by_voter_weight
...
+  def self.duplicates
+    self.approved_by_admin.select(:nominee,:role,:award_id,:theater,:show).group(:nominee,:role,:award_id,:theater,:show).select("count(*) AS count").having("count(*) > 1")
```
Comment (sumeetjain) [inline app/models/nomination.rb:34]: Line is too long. [172/80]

File: app/models/nomination.rb (hunk 49 above)
Comment (sumeetjain) [inline app/models/nomination.rb:38]: Assignment Branch Condition size for ranked_by_weight is too high. [16.28/15]

File: app/models/nomination.rb (hunk 50 above)
Comment (sumeetjain) [inline app/models/nomination.rb:48]: Space missing to the left of {.

Comment (sumeetjain) [inline app/models/nomination.rb:48]: Space between { and | missing.

Comment (sumeetjain) [inline app/models/nomination.rb:48]: Unused block argument - `k`. If it's necessary, use `_` or `_k` as an argument name to indicate that it won't be used.

Comment (sumeetjain) [inline app/models/nomination.rb:48]: Space missing after comma.

Comment (sumeetjain) [inline app/models/nomination.rb:48]: Space missing inside }.

**Pull request merged**